sw/device/silicon_creator/lib/base/sec_mmio.h

```diff
@@ -0,0 +1,189 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_SILICON_CREATOR_LIB_BASE_SEC_MMIO_H_
+#define OPENTITAN_SW_DEVICE_SILICON_CREATOR_LIB_BASE_SEC_MMIO_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @file
+ * @brief Secure Memory-mapped IO functions, for volatile access.
+ *
+ * This module is responsible for tracking critical register values for an
+ * initialized context `sec_mmio_ctx_t`, and provides a mechanism to evaluate
+ * expectations and trigger shutdown escalation on fault detection.
+ *
+ * Initialization
+ *
+ * - `sec_mmio_init()`.
+ *
+ * Register writes
+ *
+ * - Perform a number (N) of calls to `sec_mmio_write32()`.
+ * - Increment the expected number of writes by N by calling
+ *   `sec_mmio_write_increment()`. This is done using a separate function call
+ *   to be able to detect skip instruciton faults on `sec_mmio_write32()`
+ *   calls.
+ *
+ * Register reads
+ *
+ * Use the `sec_mmio_read32()`.
+ *
+ * Expectation checks
+ *
+ * See the following:
+ *
+ * - `sec_mmio_check_values()`
+ * - `sec_mmio_check_counters()`
+ *
+ * Opens:
+ *
+ * - sec_mmio_ctx is currently defined as an extern to simplify testing. For the
+ *   actual target, we need to define a memory region to share the data with the
+ *   ROM_EXT.
+ * - Currently fault detection escalations are performed by calling a handler
+ *   that is registered at `sec_mmio_init()` call time. Need to determine if we
+ *   want to move to a mock_shutdown implementation, or if we want to refactor
+ *   the code to return error codes.
+ */
+
+enum {
+  /**
+   * Number of registers stored in the sec_mmio context.
+   */
+  kSecMmioRegFileSize = 100,
```

Comment:
I assume this is just an initial estimate.  Will we tune this later?

Reply:
Done. Also added an issue to track this.